pkg/controllers/federatedhpa/util.go

```diff
@@ -0,0 +1,417 @@
+	if err = runtime.DefaultUnstructuredConverter.FromUnstructured(fieldValMap, &targetResource); err != nil {
+		return Resource{}, err
+	}
+
+	gv, err := schema.ParseGroupVersion(targetResource.APIVersion)
+	if err != nil {
+		return Resource{}, err
+	}
+
+	return Resource{
+		name:      targetResource.Name,
+		gvk:       gv.WithKind(targetResource.Kind),
+		namespace: uns.GetNamespace(),
+	}, nil
+}
+
+func getPropagationPolicyResourceFromFedWorkload(workload fedcorev1a1.GenericFederatedObject) *Resource {
+	if policyName, exists := workload.GetLabels()[scheduler.PropagationPolicyNameLabel]; exists {
+		return &Resource{
+			gvk: schema.GroupVersionKind{
+				Group:   fedcorev1a1.SchemeGroupVersion.Group,
+				Version: fedcorev1a1.SchemeGroupVersion.Version,
+				Kind:    PropagationPolicyKind,
+			},
+			name:      policyName,
+			namespace: workload.GetNamespace(),
+		}
+	}
+
+	if policyName, exists := workload.GetLabels()[scheduler.ClusterPropagationPolicyNameLabel]; exists {
+		return &Resource{
+			gvk: schema.GroupVersionKind{
+				Group:   fedcorev1a1.SchemeGroupVersion.Group,
+				Version: fedcorev1a1.SchemeGroupVersion.Version,
+				Kind:    ClusterPropagationPolicyKind,
+			},
+			name: policyName,
+		}
+	}
+
+	return nil
+}
+
+func addFedHPAEnableLabel(ctx context.Context, uns *unstructured.Unstructured) bool {
+	logger := klog.FromContext(ctx)
+
+	labels := uns.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string, 1)
+	}
+	if oldValue, ok := labels[common.FedHPAEnableKey]; ok && oldValue == common.AnnotationValueTrue {
+		return false
+	}
+
+	logger.V(3).Info("Adding fed hpa enable label")
+	labels[common.FedHPAEnableKey] = common.AnnotationValueTrue
+	uns.SetLabels(labels)
+
+	return true
+}
+
+func removeFedHPAEnableLabel(ctx context.Context, uns *unstructured.Unstructured) bool {
+	logger := klog.FromContext(ctx)
+
+	labels := uns.GetLabels()
+	if _, ok := labels[common.FedHPAEnableKey]; !ok {
+		return false
+	}
+
+	logger.V(3).Info("Removing fed hpa enable label")
+	delete(labels, common.FedHPAEnableKey)
+	uns.SetLabels(labels)
+
+	return true
+}
+
+func addFedHPAPendingController(
+	ctx context.Context,
+	fedObject fedcorev1a1.GenericFederatedObject,
+	ftc *fedcorev1a1.FederatedTypeConfig,
+) (bool, error) {
+	logger := klog.FromContext(ctx)
+
+	pendControllers, err := pendingcontrollers.GetPendingControllers(fedObject)
+	if err != nil {
+		logger.Error(err, "Failed to get pending controllers")
+		return false, err
+	}
+
+	for _, controllers := range pendControllers {
+		for _, controller := range controllers {
+			if controller == PrefixedFederatedHPAControllerName {
+				return false, nil
+			}
+		}
+	}
+
+	logger.V(3).Info(fmt.Sprintf("Setting pending controllers %v", ftc.GetControllers()))
+
+	// TODO: By default, fed-hpa controller is the first controller.
+	// 	Otherwise, this code needs to be modified.
+	_, err = pendingcontrollers.SetPendingControllers(fedObject, ftc.GetControllers())
```

Comment:
Could you reminder me how we ensure that fed-hpa-controller is the first controller?

Reply:
My idea is that we establish rules, with the fed-hpa-controller the first controller, so currently the order is as follows.
    - - kubeadmiral.io/fed-hpa-controller
    - - kubeadmiral.io/global-scheduler
    - - kubeadmiral.io/overridepolicy-controller

Reply:
It should be followed when creating hpa ftc.